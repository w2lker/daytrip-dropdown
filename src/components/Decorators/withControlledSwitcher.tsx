import React, { useState } from "react";

export interface IWithControlledSwitcher {
    selected?: string;
    onSelect?: (val: string) => void;
}

function withControlledSwitcher<Props>(Child: React.ComponentType<Props>):React.FC<Props & IWithControlledSwitcher> {
    const ControlledSwitcherProvider: React.FC<Props> = (props) => {
        const {selected, onSelect, ...other} = props as IWithControlledSwitcher;
        const [controlledValue, setControlledValue] = useState('');
        if (selected && onSelect) {
            return <Child {...props} />;
        }
        const setControlledValueWithCallback = (selected: string) => {
            setControlledValue(selected);
            onSelect && onSelect(selected);
        };
        // @ts-ignore
        return <Child {...other} selected={controlledValue} onSelect={setControlledValueWithCallback}/>
    };
    ControlledSwitcherProvider.displayName = `WithControlledSwitcher(${Child.displayName})`;
    return ControlledSwitcherProvider;
}

export default withControlledSwitcher;
