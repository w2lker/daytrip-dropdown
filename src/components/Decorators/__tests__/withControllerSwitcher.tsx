import React, { ReactElement } from "react";
import {mount, shallow} from "enzyme";
import withControlledSwitcher from "../withControlledSwitcher";

const setup = () => {
  const SampleComponent = ():ReactElement => {
    return (<></>);
  };
  const sampleProps = {
    customProp: 'customProp-passes',
    select: 'someValue',
    onSelect: jest.fn()
  };
  return {SampleComponent, sampleProps};
};

it('match snapshot', () => {
  const {SampleComponent, sampleProps} = setup();
  const DecoratedComponent = withControlledSwitcher(SampleComponent);
  const component = mount(<DecoratedComponent {...sampleProps}/>);
  expect(component.debug()).toMatchSnapshot();
});

it('pass original props if they are provided', () => {
  const {SampleComponent, sampleProps} = setup();
  const DecoratedComponent = withControlledSwitcher(SampleComponent);
  const component = shallow(<DecoratedComponent {...sampleProps}/>);
  const props = component.find(SampleComponent).props();
  // @ts-ignore
  expect(props.customProp).toBe(sampleProps.customProp);
  // @ts-ignore
  props.onSelect();
  expect(sampleProps.onSelect).toBeCalledTimes(1);
});

it('provide state controller if non of {value, onSelect} provided', () => {
  const {SampleComponent} = setup();
  const DecoratedComponent = withControlledSwitcher(SampleComponent);
  const component = shallow(<DecoratedComponent />);
  const props = component.find(SampleComponent).props();
  // @ts-ignore
  expect(props.onSelect).toBeDefined();
  // @ts-ignore
  expect(typeof props.onSelect).toBe('function');
});

it('fires callback if value is not provided', () => {
  const {SampleComponent, sampleProps} = setup();
  const noValProps = {
    ...sampleProps,
    value: undefined,
  };

  const DecoratedComponent = withControlledSwitcher(SampleComponent);
  const component = shallow(<DecoratedComponent {...noValProps}/>);
  const props = component.find(SampleComponent).props();
  const { onSelect } = props as {onSelect: (val: string) => void};
  expect(onSelect).toBeDefined();
  expect(typeof onSelect).toBe('function');
  onSelect('some-val');
  expect(sampleProps.onSelect).toBeCalledTimes(1);
  expect(sampleProps.onSelect.mock.calls[0][0]).toBe('some-val');
});
