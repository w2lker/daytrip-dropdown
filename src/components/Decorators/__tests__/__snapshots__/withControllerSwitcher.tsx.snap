// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match snapshot 1`] = `
"<WithControlledSwitcher(undefined) customProp=\\"customProp-passes\\" select=\\"someValue\\" onSelect={[Function: mockConstructor]}>
  <SampleComponent customProp=\\"customProp-passes\\" select=\\"someValue\\" selected=\\"\\" onSelect={[Function: setControlledValueWithCallback]} />
</WithControlledSwitcher(undefined)>"
`;
